pilosa/client.py

```diff
@@ -200,30 +201,37 @@ def _import_bits(self, index_name, frame_name, slice, bits):
             client._import_node(_ImportRequest(index_name, frame_name, slice, bits))
 
     def _fetch_fragment_nodes(self, index_name, slice):
-        uri = "%s/fragment/nodes?slice=%d&index=%s" % (self.__get_address(), slice, index_name)
-        response = self.__http_request("GET", uri, client_response=Client.__ERROR_CHECKED_RESPONSE)
+        path = "/fragment/nodes?slice=%d&index=%s" % (slice, index_name)
+        response = self.__http_request("GET", path, client_response=Client.__ERROR_CHECKED_RESPONSE)
         content = response.data.decode('utf-8')
         return json.loads(content)
 
     def _import_node(self, import_request):
         data = import_request.to_protobuf()
-        uri = "%s/import" % self.__get_address()
-        self.__http_request("POST", uri, data=data)
+        self.__http_request("POST", "/import", data=data, client_response=Client.__RAW_RESPONSE)
 
     def __patch_index_time_quantum(self, index):
-        uri = "%s/index/%s/time-quantum" % (self.__get_address(), index.name)
+        path = "/index/%s/time-quantum" % index.name
         data = '{\"timeQuantum\":\"%s\"}"' % str(index.time_quantum)
-        self.__http_request("PATCH", uri, data=data)
+        self.__http_request("PATCH", path, data=data)
 
-    def __http_request(self, method, uri, data=None, client_response=0):
+    def __http_request(self, method, path, data=None, client_response=0):
         if not self.__client:
             self.__connect()
-        try:
-            response = self.__client.request(method, uri, body=data)
-        except urllib3.exceptions.MaxRetryError as e:
-            self.cluster.remove_host(self.__current_host)
-            self.__current_host = None
-            raise PilosaError(str(e))
+        # try at most 10 non-failed hosts; protect against broken cluster.remove_host
+        for _ in range(_MAX_HOSTS):
+            uri = "%s%s" % (self.__get_address(), path)
+            try:
+                _LOGGER.debug("Request: %s %s %s", method, uri,
+                              "[binary]" if client_response == Client.__RAW_RESPONSE else data)
+                response = self.__client.request(method, uri, body=data)
+                break
+            except urllib3.exceptions.MaxRetryError as e:
+                self.cluster.remove_host(self.__current_host)
```

Comment:
I don't think user intervention should be required to "reset" the cluster once all hosts are blacklisted. The client should make its best effort to perform the request - if connections to pilosa hosts are persistently failing, it should return an error to the calling code. If the calling code wants to keep trying, it can re-issue the request, and the client can try connecting to the cluster again.

Changing the internal state of the client between requests, and adding to the API to help calling code manage that state feels like more complexity than is needed at this point.